src/keri/vc/protocoling.py

```diff
@@ -405,3 +405,221 @@ def presentationExchangeExn(hab, reger, said):
+            **kwa (dict): keyword arguments passed to DoDoer
+
+        """
+        self.hby = hby
+        self.rgy = rgy
+        self.notifier = notifier
+        self.msgs = decking.Deck()
+        self.cues = decking.Deck()
+
+        super(IssueHandler, self).__init__(**kwa)
+
+    def recur(self, tyme):
+        if self.msgs:
+            msg = self.msgs.popleft()
+            serder = msg["serder"]
+            attrs = serder.ked["a"]
+
+            if not attrs["r"].startWith(self.resource):
+                return False
+            data = dict(
+                r=f"/exn{attrs['r']}",
+                d=serder.said,
+                m=attrs["m"]
+            )
+
+            self.notifier.add(attrs=data)
+
+        return False
+    
+def ipexApplyExn(hab, message, schema, attrs):
+    """ Apply for an ACDC
+
+    Parameters:
+        hab(Hab): identifier environment for issuer of credential
+        message(str): Human readable message regarding the credential application
+        schema (any): schema or its SAID
+        attrs (any): attribute field label list
+
+    Returns:
+        Serder: credential issuance exn peer to peer message
+        bytes: attachments for exn message
+
+    """
+    data = dict(
+        m=message,
+        s=schema,
+        a=attrs
+    )
+
+    exn, end = exchanging.exchange(route="/ipex/apply", payload=data, sender=hab.pre)
+    ims = hab.endorse(serder=exn, last=False, pipelined=False)
+    del ims[:exn.size]
+    ims.extend(end)
+
+    return exn, ims
+
+def ipexOfferExn(hab, message, offer):
+    """ Offer a metadata ACDC
+
+    Parameters:
+        hab(Hab): identifier environment for issuer of credential
+        message(str): Human readable message regarding the credential offer
+        offer (any): metadata ACDC or its SAID
+
+    Returns:
+        Serder: credential issuance exn peer to peer message
+        bytes: attachments for exn message
+
+    """
+    data = dict(
+        m=message,
+        o=offer,
+    )
+
+    exn, end = exchanging.exchange(route="/ipex/offer", payload=data, sender=hab.pre)
+    ims = hab.endorse(serder=exn, last=False, pipelined=False)
+    del ims[:exn.size]
+    ims.extend(end)
+
+    return exn, ims
+
+def ipexAgreeExn(hab, message, offer):
+    """ Agree an offer
+
+    Parameters:
+        hab(Hab): identifier environment for issuer of credential
+        message(str): Human readable message regarding the credential agreement
+        offer (any): offer received or its SAID
+
+    Returns:
+        Serder: credential issuance exn peer to peer message
+        bytes: attachments for exn message
+
+    """
+    data = dict(
+        m=message,
+        o=offer
+    )
+
+    exn, end = exchanging.exchange(route="/ipex/agree", payload=data, sender=hab.pre)
+    ims = hab.endorse(serder=exn, last=False, pipelined=False)
+    del ims[:exn.size]
+    ims.extend(end)
+
+    return exn, ims
+    
+def ipexGrantExn(hab, message, acdc, iss):
```

Comment:
OK, that can be an `ixn` or a `rot`. Do you have a var name that covers both type of events or is it better to include the two and make them optional?

Reply:
It doesn't matter to the protocol whether it's an ixn or rot so no need to make the caller choose. Let's go with anc for anchoring event and then it can be either. This avoids optional values in the exn.  

This isn't the approach I took with registry creation but it should be. I'll update that.

Reply:
`anc` added